packages/csv-parser-state/src/main.js

```diff
@@ -0,0 +1,190 @@
+    {
+      apiConfig,
+      csvConfig,
+      continueOnProblems = false,
+      accessToken,
+    }: ConstructorOptions,
+    logger: LoggerOptions
+  ) {
+    this.apiConfig = apiConfig
+    this.accessToken = accessToken
+
+    const defaultOptions = {
+      delimiter: ',',
+      multiValueDelimiter: ';',
+    }
+    this.csvConfig = { ...defaultOptions, ...csvConfig }
+    this.continueOnProblems = continueOnProblems
+    this.logger = logger || {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      debug: () => {},
+    }
+
+    this._rowIndex = 0
+  }
+
+  parse(input: stream$Readable, output: stream$Writable) {
+    this.logger.info('Starting conversion')
+    highland(input)
+      .through(csv({ separator: this.csvConfig.delimiter, strict: true }))
+      .map(CsvParserState._removeEmptyFields)
+      .map(CsvParserState._parseInitialToBoolean)
+      .map(unflatten)
+      .map(state => this._mapMultiValueFieldsToArray(state))
+      .flatMap(state => highland(this._transformTransitions(state)))
+      .errors((error, cb) => this._handleErrors(error, cb))
+      .stopOnError(error => {
+        // <- Emit error and close stream if needed
+        this.logger.error(`At row: ${this._rowIndex}, ${error}`)
+        output.emit('error', error)
+      })
+      .doto(() => {
+        this._rowIndex += 1
+        this.logger.debug(`At row: ${this._rowIndex}, Successfully parsed`)
+      })
+      .pipe(JSONStream.stringify())
+      .pipe(output)
+  }
+
+  async _transformTransitions({
+    transitions,
+    ...restOfState
+  }: StateWithUnresolvedTransitions) {
+    if (!transitions) return restOfState
+    // We setup the client here because it is unnecessary
+    // if there are no transitions
+    this._setupClient()
+    const stateRequests = transitions.map(transitionState =>
+      this._buildStateRequest(transitionState)
+    )
+    const resolvedStates = await Promise.all(stateRequests)
+    const stateReferences = resolvedStates.map(response => ({
+      typeId: 'state',
+      id: response.body.id,
+    }))
+    return { ...restOfState, transitions: stateReferences }
+  }
+
+  _buildStateRequest(stateKey: string): Promise<SuccessResult> {
+    const stateService = this._createStateService()
+    const uri = stateService.byKey(stateKey).build()
+    return this._fetchStates(uri)
+  }
+
+  _setupClient() {
+    if (!this.apiConfig)
+      throw new Error('The constructor must be passed an `apiConfig` object')
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          this.accessToken ? `Bearer ${this.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+  }
+
+  _createStateService() {
```

Comment:
I think this can be moved out of the class.

```js
const createCow = () => {}

class CowFarm {
  super() {
    this.cow = createCow()
  }
}
```

Reply:
Or maybe a static method would be better